### PR TITLE
Ben/sequence multi labeling

### PR DESCRIPTION
**Pull request by benleetownsend**

### CODE REVIEW

**Comment by benleetownsend** [comment]
This currently breaks regular sequence labeling

File: finetune/reweighting.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
+import tensorflow as tf
+
+
+def learning_to_reweight(gold_data, gold_targets, data, targets, model, loss, lr=1e-5):
```
Comment (madisonmay) [inline finetune/reweighting.py]: This should be cut from this PR.

File: finetune/reweighting_classifier.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
+import tensorflow as tf
+import numpy as np
+
+from finetune.base import BaseModel
+from finetune.target_encoders import OneHotLabelEncoder
+from finetune.network_modules import classifier
+
+
+class ReweightingClassifier(BaseModel):
```
Comment (madisonmay) [inline finetune/reweighting_classifier.py]: Cut this from this PR.

File: finetune/sequence_labeling.py [hunk 3]
```diff
@@ -155,10 +161,13 @@ def predict(self, X, max_length=None):
                     # format probabilities as dictionary
                     probs = np.mean(np.vstack(prob_seq), axis=0)
                     prob_dicts.append(dict(zip(self.label_encoder.classes_, probs)))
+                    if self.multi_label:
+                        del prob_dicts[-1][self.config.pad_token]
                 
                 all_subseqs.append(doc_subseqs)
                 all_labels.append(doc_labels)
                 all_probs.append(prob_dicts)
+        print(all_labels[0])
```
Comment (madisonmay) [inline finetune/sequence_labeling.py]: print

File: finetune/sequence_labeling.py [hunk 4]
```diff
@@ -220,4 +231,31 @@ def _predict_op(self, logits, **kwargs):
         return label_idxs, label_probas
 
     def _predict_proba_op(self, logits, **kwargs):
-        return tf.no_op()
\ No newline at end of file
+        return tf.no_op()
+
+
+class SequenceMultiLabeler(SequenceLabeler):
```
Comment (madisonmay) [inline finetune/sequence_labeling.py]: Odd class name but I don't have a better suggestion at the moment.

Comment (madisonmay) [reply finetune/sequence_labeling.py]: I think we talked about making this a config argument rather than a class.  Looks like you've already done that for the most part.

File: finetune/utils.py [hunk 5]
```diff
@@ -291,72 +297,85 @@ def finetune_to_indico_sequence(raw_texts, subseqs, labels, probs=None, none_val
...
+                label_list = raw_label
+
+            for label in label_list:
+
+                stripped_text = sub_str.strip()
+
+                raw_annotation_start = raw_text.find(stripped_text, raw_annotation_start)
+                for i, item in enumerate(doc_annotations):
+                    if item["label"] == label and item["end"] == raw_annotation_end:
+                        doc_annotations[-1], doc_annotations[i] = doc_annotations[i], doc_annotations[-1]
```
Comment (madisonmay) [inline finetune/utils.py]: I don't understand the swapping.

File: finetune/utils.py [hunk 6]
```diff
@@ -376,23 +395,58 @@ def indico_to_finetune_sequence(texts, labels=None, none_value=config.PAD_TOKEN)
...
+                if multi_label:
+                    doc_labels.append([none_value])
+                else:
+                    doc_labels.append(none_value)
+
+            if start < last_loc:
+                if not multi_label:
+                    raise ValueError("Overlapping annotations requires the multi-label model")
+                else:
+                    prev_end = label_seq[i - 1]["end"]
```
Comment (madisonmay) [inline finetune/utils.py]: why are these set to the same thing?

Comment (madisonmay) [reply finetune/utils.py]: Tests to write:
Overlapping
Label contained within another
More than 2 labels

**Comment by benleetownsend** [comment]
Im just running tests after merging the two sequence labelers.
